vm_insnhelper.c

```diff
@@ -6204,18 +6204,38 @@ VALUE rb_setup_fake_ary(struct RArray *fake_ary, const VALUE *list, long len, bo
 VALUE rb_ec_pack_ary(rb_execution_context_t *ec, VALUE ary, VALUE fmt, VALUE buffer);
 
 VALUE
-rb_vm_opt_newarray_pack(rb_execution_context_t *ec, rb_num_t num, const VALUE *ptr, VALUE fmt)
+rb_vm_opt_newarray_pack_buffer(rb_execution_context_t *ec, rb_num_t num, const VALUE *ptr, VALUE fmt, VALUE buffer)
 {
     if (BASIC_OP_UNREDEFINED_P(BOP_PACK, ARRAY_REDEFINED_OP_FLAG)) {
         struct RArray fake_ary;
         VALUE ary = rb_setup_fake_ary(&fake_ary, ptr, num, true);
-        return rb_ec_pack_ary(ec, ary, fmt, Qnil);
+        return rb_ec_pack_ary(ec, ary, fmt, buffer);
     }
     else {
-        return rb_vm_call_with_refinements(ec, rb_ary_new4(num, ptr), idPack, 1, &fmt, RB_PASS_CALLED_KEYWORDS);
+        // The opt_newarray_send insn drops the keyword args so we need to rebuild them.
+        // Setup an array with room for keyword hash.
+        VALUE args[2];
+        args[0] = fmt;
+        int kw_splat = RB_NO_KEYWORDS;
+        int argc = 1;
+
+        if (!NIL_P(buffer)) {
```

Comment:
This have it not being able to distinguish between passing `buffer: nil` and not passing a keyword at all. The following raises when it should not:

```
class Array
  def pack(_fmt, buffer:) = buffer
end

v = 1
b = nil;
p [v].pack("E*", buffer: b)
```

Reply:
Indeed, thank you!
At one point I was trying to figure out if it would be odd for the method to receive the keyword arg when it wasn't provided (something like `def pack(_fmt, **kwargs); kwargs.empty?`) and then apparently I didn't add enough tests.

Reply:
Do you think the difference is important enough to warrant having duplicate logic in `rb_vm_opt_newarray_pack`
(rather than having it just be a wrapper around `rb_vm_opt_newarray_pack_buffer`)?

Reply:
Maybe you can get it to work by passing `Qundef` to indicate that the callsite passes no keywords?

Reply:
Oh, good point.  I forgot we had that one.  Thanks!

Reply:
Switched to `Qundef` and added tests.